feat: Split creation of path structure from actual download of images…

blenzi opened the pull request:
… in OpenFire

- Add methods create_directories and check_images
- Rework method download
- Add option max_files
- Use max_files = 100 in test/test_datasets.py

Resolves: #26

frgfm committed:
Initial commit

frgfm committed:
Project setup (#1)

* docs: Updated readme

* docs: Added contribution guidelines

* feat: Added utils module

Added environment information script for issuers.

* chore: Added package setup

* docs: Updated readme

Added installation instructions

* docs: Added issue templates

* chore: Added CI

* test: Added unittests for utils

* test: Added unittests into CI

* docs: Updated readme

Added CI and coverage badges

* refactor: Removed unused import

frgfm committed:
Fixed package installation (#5)

* chore: Fixed package installation

* docs: Updated readme

Updated installation instructions

* chore: Updated package metadata

* chore: Added __version__ attribute to package

frgfm committed:
Added package documentation (#7)

* docs: Added docstring

* docs: Added Sphinx doc generation

* chore: Added doc jobs to CI

Closes #4

* docs: Updated readme

Added documentation badge and link.

x0s committed:
docs: Detail project setup and contributing workflow (#8)

* Add new table of contents to precise project setup & contributing workflow

* Add explanation on how to create a virtual environment

* Add section to detail how to fork project and install dependencies

* detail step-by-step the current contributing worklow

frgfm committed:
feat: Added datasets module (#11)

* feat: Added datasets module

Added URL download function in pyronear.datasets.utils.

* test: Added datasets.utils unittests

* chore: Added __init__ to datasets

* docs: Added datasets docs

* docs: Updated doc theme

* docs: Updated color theme

* chore: Fixed datasets __init__

* refactor: Refactored download_url

Reduced dependencies and sped up execution

* feat: Added OpenFire dataset

Inherits from torchvision.datasets.vision.VisionDataset

* style: Updated docstrings

* docs: Added OpenFire to datasets

* chore: Updated requirements.txt

* refactor: Refactored OpenFire for consistency

Removed torch.Tensor casting of target since the collate function of DataLoader will cast it anyway. Added a class_to_idx attribute. Closes #6

* test: Updated datasets.utils test

Sped up basename resolution

* test: Added unittests for datasets.OpenFire

* test: Fixed OpenFire unittests

As per #10, a few files can be unavailable for download temporarily.

* style: Improved warning for OpenFire download

Improved warning message that are thrown by OpenFire.download() to display the error that was caught.

* test: Refactored datasets unittests

* fix: Fixed datasets.OpenFire

frgfm committed:
chore: Changed CI build with bot (#13)

Switched Github account used for build to bot

frgfm committed:
feat: Updated datasets.OpenFire (#15)

* test: Moved temporary path to context

* test: Added tests for download_url

* feat: Added multi-thread URL download

See #14

* docs: Updated docstring

* refactor: Switched silent mode to optional

* refactor: Fixed get_fname

More tolerant base name resolution

* test: Added unittests for download_urls

* feat: Updated URL to cleaned OpenFire

See #6 

* feat: Added multi-thread download to OpenFire

Closes #14

* chore: Updated requirements.txt

* style: Removed whitespaces

* refactor: Removed unused import

* style: Improved exception handling

Commit:
test: Using module tempfile to avoid failed test on #15 (#16)

* test: Using module tempfile to avoid failed test on #15

Changed test/test_datasets.py according to suggestion from https://app.codacy.com/manual/fg/pyronear/pullRequest?prid=4288087

* fix: problem with usage of TemporaryDirectory on #16

Changed test/test_datasets.py

frgfm committed:
style: Renamed unittest and fixed __init__ (#17)

* test: Updated utils test name

* refactor: Removed utils scripts from __init__

* test: Updated utils import

frgfm committed:
feat: Updated OpenFire URL (#18)

* feat: Updated OpenFire URL

See #6 

* docs: Updated readme

* fix: Fixed large image loading with pillow

* feat: Added option to create validation set

A valid_pct can now be passed when download=True to create the validation set by taking a percentage of the training set.

* test: Added unittests to OpenFire

Added test for the valid_pct argument of OpenFire

* feat: Improved random split of OpenFire sets

Used local seed setting of the random module to avoid disturbing global execution.

frgfm committed:
fix: Fixed OpenFire image extension guessing (#21)

* fix: Fixed image extension for OpenFire

Since extension guessing was not verified as valid extension, some strings of digits could be taken as an extension. This resulted in images with .760 extensions which can be problematic on some OS. Closes #20

* test: Added unittest for OpenFire image extensions

See #20

* fix: Fixed extension guessing

Previous fix assumed only image types.

* test: Updated extension testing for OpenFire

frgfm committed:
feat: Added reference scripts for binary classification (#23)

* feat: Added fastai training script

Model's head still has to be switched to binary classification and ImageDataBunch needs to be built while respecting the existing split of train/validation. See #3

* feat: Updated fastai training script

Fixed the split of train and valid set so that the ImageDataBunch takes the existing split. See #3

* feat: Added training script without fastai

See #3

* feat: Added deterministic mode to torch_train

See #3

* style: Added endfile lines

* fix: Fixed training script train/valid split

For users that did not have the validation split, the training was failing when instantiating OpenFire in non training mode.

* refactor: Reorganized reference scripts

Separated fastai and non-fastai scripts to be able to have different requirements.txt for cleaner usage.

* docs: Updated readme

Added instructions for training using references scripts.

* refactor: Removed unused imports and variables

* fix: Fixed fastai training script

Missing pandas dependency

* fix: Fixed scheduler step on training scripts

* fix: Reflected script argument in scheduler

* refactor: Removed unused variable

* feat: Updated torch training script

Changed evaluation and logging to have similar metrics to fastai script. See #3

* chore: Updated version

Switched from alpha to beta.

* refactor: Harmonize default data path between scripts

* chore: Updated training script requirements

OneCycleLR is only available since torch 1.3.0.

* style: Renamed learner

* feat: Added device setting to fastai training

* refactor: Unified device resolution for training scripts

Closes #3 
See #19, #24

Commit:
feat: Split creation of path structure from actual download of images in OpenFire

- Add methods create_directories and check_images
- Rework method download
- Add option max_files
- Use max_files = 100 in test/test_datasets.py

Resolves: #26

Commit:
feat: Split creation of path structure from actual download of images in OpenFire

- Add methods create_directories and check_images
- Rework method download
- Add option max_files
- Use max_files = 100 in test/test_datasets.py

Resolves: #26

Commit:
fix: fix previous commit

Remove unused variable in openfire.py OpenFire.check_images
Remove trailing whitespaces
Fix warning message for samples that could not be downloaded

Commit:
fix: fix previous commit

Remove unused variable in openfire.py OpenFire.check_images
Remove trailing whitespaces
Fix warning message for samples that could not be downloaded

Commit:
fix: lines added by mistake in previous commit

pyronear/datasets/openfire.py:
- Remove lines added by mistake in previous commit
- Remove trailing whitespaces

Commit:
fix: lines added by mistake in previous commit

pyronear/datasets/openfire.py:
- Remove lines added by mistake in previous commit
- Remove trailing whitespaces

Commit:
fix: remove trailing whitespaces in pyronear/datasets/openfire.py

Commit:
fix: remove trailing whitespaces in pyronear/datasets/openfire.py

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -17,6 +17,18 @@
+    | - OpenFire/
+    | | - processed/
+    | | |  - test.pl
+    | | |  - training.pl
```
Little typo here: `.pt` extensions, not `.pl`

blenzi replied:
Thanks, fixed

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -42,13 +54,16 @@ class OpenFire(VisionDataset):
+          self.create_directories(max_files=max_files)
+          if download:
+              self.download(threads)
+          self.check_images(valid_pct=valid_pct)
```
Spacing issue here, one tab should be equal to 4 spaces in your editor (rather than 2 here most likely)

blenzi replied:
Fixed

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -42,13 +54,16 @@ class OpenFire(VisionDataset):
-        if download:
-            self.download(threads, valid_pct)
+        
+        if not (self._check_exists(train=True) and self._check_exists(train=False)):
```
Not sure to understand the logic behind this line.
The lines below will be executed only if test set exists already but not the training one?

blenzi replied:
They will be skipped if both exist. Maybe you missed the () ?
(N.B.: this will disappear if we create the .pt files each time)

frgfm replied:
Indeed, my bad. Better to go for  `not A or not B` (it will stop earlier if the first condition is met, whereas in your suggestion both will be evaluated)

blenzi replied:
Aesthetically I agree but I believe both are equivalent. The following does not execute the 2nd statement:

```
import time
if not (False and time.sleep(5)): pass
```

frgfm replied:
I agree but I was referring to execution time, but I just tested it and it doesn't seem to be doing much difference indeed so that's alright!

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,55 @@ def _check_exists(self, train=True):
-        training_set, test_set = [], []
+        # Add the local filename to the annotations
+        for idx, a in enumerate(self.annotations):
+          a['fname'] = f"{idx:06}.{get_fname(a['url']).rpartition('.')[-1]}"
```
Spacing issue here (should be 4 spaces rather than 2)

blenzi replied:
Fixed

frgfm reviewed (changes requested):
While I understand the need to split:
- downloading the images
- verifying the images and saving the `.pt` extracts

There are two additional features that are being proposed here:

1. Separating hierarchy creation & download
I don't see any reason why one would be performed without the other?

2. Setting a limited number of files in extract
For actual public use of the library (non-testing purposes), limiting the download to a subtract of the whole might be troublesome if different users unexpectedly pass different values but expected to get the same dataset. Apart from unittests, is there any other use cases?

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,55 @@ def _check_exists(self, train=True):
+        """
+        Create the directory structure to hold the images and training_file/test_file.
+        Download the json file with annotations, add 'fname' with local filenames
+        
```
Trailing whitespace here according to Codacy

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,55 @@ def _check_exists(self, train=True):
+        """
+        Create the directory structure to hold the images and training_file/test_file.
+        Download the json file with annotations, add 'fname' with local filenames
+        
```
There is a trailing whitespace here according to Codacy.

blenzi replied:
Fixed (codacy is annoying, this is just documentation)

blenzi commented:
Thanks for the review!

> While I understand the need to split:
> 
>     * downloading the images
> 
>     * verifying the images and saving the `.pt` extracts
> 
> 
> There are two additional features that are being proposed here:
> 
>     1. Separating hierarchy creation & download
>        I don't see any reason why one would be performed without the other?

It was to prepare the next PR in which the user would be able to pass an existing image folder, to be linked to the one used by OpenFire.

Maybe I can ask differently: why this structure is needed ? Would it be enough to pass an image folder, which is downloaded if not given?

Then the code could store the .pt files used for training and testing in pytorch in a temporary location. This is fast and avoids the risk of a user removing the images while keeping the .pt files (I have done it accidentally already...).

And in the future, using other datasets could be done easily by passing another json file as argument as well. Maybe this is what you had in mind ?

> 
>     2. Setting a limited number of files in extract
>        For actual public use of the library (non-testing purposes), limiting the download to a subtract of the whole might be troublesome if different users unexpectedly pass different values but expected to get the same dataset. Apart from unittests, is there any other use cases?

Not that I can think of but I believe it is an important use case. Adding a comment that the parameter should only be changed for testing purposes would be enough in your opinion ?

frgfm commented:
> It was to prepare the next PR in which the user would be able to pass an existing image folder, to be linked to the one used by OpenFire.
I fully agree, but even if the user passes a dataset, the folder hierarchy will have to be verified, or am I missing something?
Probably having something similar to 
```
if not my_folder.is_folder():
    my_folder.mkdir(parents=True)
```
to leverage the use of `pathlib.Path` objects.

> Maybe I can ask differently: why this structure is needed ? Would it be enough to pass an image folder, which is downloaded if not given?

Good point, I thought about it more as a convention where all the libraries downloading datasets can share a same root (hence the default `./data` value), where all their datasets are stored. PyTorch does it both for datasets and models. 
The user can easily free some memory if she/he wants to delete the downloaded data at once (or move it for that matter). And after that, expectedly, you have to split the data within the root depending on its project/source.

> Then the code could store the .pt files used for training and testing in pytorch in a temporary location. This is fast and avoids the risk of a user removing the images while keeping the .pt files (I have done it accidentally already...).

Good idea, fully agree with you on that one.

> And in the future, using other datasets could be done easily by passing another json file as argument as well. Maybe this is what you had in mind ?

Yes, the core idea is that the  ̀OpenFire` dataset will either be split into `OpenFireBin` `OpenFireMulti`, etc for the different tasks, or take an argument `task` (preferably). Based on the value of `task` it will either fetch the  ̀.json` from a different URL (difference in annotations), or download everything at once, and return the appropriate version of the annotated dataset (my suggestion).

> Not that I can think of but I believe it is an important use case. Adding a comment that the parameter should only be changed for testing purposes would be enough in your opinion?

My point is more that exposing methods/functions for users while the feature is only used by developers (actually code maintainers, not necessarily developers) and confusing for users, might not be our best option here. Ideally, the operation of limiting the extract will be only available and performed within the test, but I agree it's not easily done. A good lead might be to take a look at https://github.com/pytorch/vision/blob/master/test/test_datasets.py#L84-L91

That being said, if the point is about speeding up the CI job, in comparison, we are still at a very reasonable duration compared to `torchvision` or other similar libraries.

Commit:
style: fixing whitespaces, indentation and typo in the documentation

Implementing fixes suggested by frgfm on #28

ecalmon-cms committed:
style: fixing whitespaces, indentation and typo in the documentation

Implementing fixes suggested by frgfm on #28

ecalmon-cms committed:
style: fixing whitespaces again...

Commit:
style: fixing whitespaces again...

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,55 @@ def _check_exists(self, train=True):
 
-    def download(self, threads=None, valid_pct=None):
-        """Download the OpenFire data if it doesn't exist in processed_folder already.
+    def create_directories(self, max_files=None):
```
default value of `max_files` differs from line 57, is it normal?

blenzi replied:
Thanks for catching it. It should be None there as well, otherwise the last file would be skipped (it was a bug on the original implementation)

frgfm replied:
Also, perhaps rename it to `num_samples` for a less confusing variable naming

blenzi replied:
Done

Commit:
fix: changing default value of max_files to None

The value was set to None in OpenFire.create_directories but not in the constructor. Setting it to -1 as before would skip the last file

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,55 @@ def _check_exists(self, train=True):
+            try:
+                self.annotations = json.load(f)[:max_files]
+            except TypeError:
+                raise ValueError("max_files must an int between -1 and len(dataset)")
```
You put a `TypeError` in except but raise a `ValueError`.
I would rather go for something like:
```
self.annotations = json.load(f)
if max_files is not None:
    self.annotations = self.annotations[:max_files]
```
It will throw the error by itself if indexing type/value isn't correct

However, I don't get why negative values are allowed considering the feature is limiting data samples (not indexing them)

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -42,13 +54,16 @@ class OpenFire(VisionDataset):
 
     def __init__(self, root, train=True, transform=None, target_transform=None,
-                 download=False, threads=16, valid_pct=None):
+                 download=False, threads=16, valid_pct=0.2, max_files=-1):
```
default value of `max_files` means that the extract will not take the last sample. I'm not sure it is expected as a behaviour! The default value should produce the same output as before the feature

Commit:
style: changing max_files to num_samples as suggested in #28

Changing new parameter name from max_files to num_samples

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -17,6 +17,18 @@
+    | | |  - training.pt
+    | | - raw/
+    | | | - images/
+
```
No need for double line break here

Commit:
style: changing max_files to num_samples as suggested in #28

Changing new parameter name from max_files to num_samples

Commit:
fix: changing max_files to num_samples in test_datasets.py as suggested in #28

Changing new parameter name from max_files to num_samples

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,54 @@ def _check_exists(self, train=True):
         download_url(self.url, self._root.joinpath(self._raw), filename=self.url.rpartition('/')[-1], verbose=False)
         with open(self._root.joinpath(self._raw, self.url.rpartition('/')[-1]), 'rb') as f:
-            annotations = json.load(f)
+            try:
+                self.annotations = json.load(f)[:num_samples]
+            except TypeError:
+                raise ValueError("num_samples must an int between -1 and len(dataset)")
```
Better to go for
```
self.annotations = json.load(f)
if num_samples is not None:
    self.annotations = self.annotations[:num_samples]
```
It will throw the appropriate error by itself if the indexing is incorrect

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -101,38 +116,54 @@ def _check_exists(self, train=True):
-        training_set, test_set = [], []
+        # Add the local filename to the annotations
+        for idx, a in enumerate(self.annotations):
+            a['fname'] = f"{idx:06}.{get_fname(a['url']).rpartition('.')[-1]}"
```
```
self.annotations[idx]['fname'] = ...
```
would be more appropriate for clear inplace modification of class attribute

frgfm commented on pyronear/datasets/openfire.py:
```diff
@@ -42,13 +54,16 @@ class OpenFire(VisionDataset):
 
-        if download:
-            self.download(threads, valid_pct)
+        if not (self._check_exists(train=True) and self._check_exists(train=False)):
+            self.create_directories(num_samples=num_samples)
+            if download:
+                self.download(threads)
+            self.check_images(valid_pct=valid_pct)
```
We have to think a bit much about this:
if you provide a different value of  `valid_pct` with `download=True`, currently, the condition of  `self._check_exists` will not be verified. So it will use extract that was created with a different value of `valid_pct`

frgfm reviewed (changes requested):
A few changes to make for the feature to be compatible and have a good expected behaviour.
The feature proposes to split the following operations:
- folder hierarchy creation
- sample downloading
- on-the-fly extract saving with custom train/valid split

Here would be the wanted behaviour:
- if  `download=False`: the class will try to use existing saved extracts (throwing an error otherwise like previously)
- if `download=True`: 
1. folder hierarchy will be created if it does not exist yet.
2. the JSON extract will be download if it does not exist yet
3. the actual images will be downloaded (if `num_samples` is provided, only the first `num_samples` will be downloaded)
4. samples will be verified (throwing a warning if the specified amount of samples has not been downloaded or is not available)

- Whatever  the value of `download`:
5. `valid_pct` of total available samples will used for validation and the rest for training

If `download=True`:
- steps 1, 2 can be skipped
- step 3 cannot (internally it will skip download if the file already exists)
- step 4 cannot
Extract saving at this point might be irrelevant
- step 5 is always executed

blenzi commented:
Thanks for the in-depth review of the logic. While I agree with most of the comments, I believe we are missing the reasons why this issue/PR was open:

- Downloading the entire dataset is not fast and is essentially never succesful. The user is forced to either manually create a folder structure that was not explicitly documented or interrupt the execution and copy 3 GB into a folder inside a nested structure. This is not convenient, in particular when using a system that requires starting from scratch like google colab.

- Without limiting the number of files, no quick tests of new functionalities or new training methods are possible (at least easily).

- The class was designed to be interfaced with torch directly while working with fastai requires quite a few workarounds. The train/test split and the .pt files are not particularly relevant.

The suggestion would be to add optional arguments:

- num_samples (already implemented) 
- image_folder, to be linked to OpenFire/raw/images (instead of copying)
- Allow for train=None, 'all' or change the argument name, in order to get the full dataset
- Add method to_ImageList for easy use with fastai

Also:

>     * Whatever  the value of `download`:
> 
> 
>     1. `valid_pct` of total available samples will used for validation and the rest for training

The current behaviour is to use the existing train/test splitting if download is False. Is this what is wanted ?

Maybe it's easier to drop this PR and restart from a fresh one if you agree with the points above

frgfm commented:
I agree @blenzi 
Let's decide the design specs on the issue #26 and work on a clean PR for it!
